measureme/src/file_header.rs

```diff
@@ -0,0 +1,95 @@
+//! All binary files generated by measureme have a simple file header that
+//! consists of a 4 byte file magic string and a 4 byte little-endian version
+//! number.
+
+use byteorder::{ByteOrder, LittleEndian};
+use crate::serialization::SerializationSink;
+use std::error::Error;
+
+pub const CURRENT_FILE_FORMAT_VERSION: u32 = 0;
+pub const FILE_MAGIC_EVENT_STREAM: &[u8; 4] = b"MMES";
+pub const FILE_MAGIC_STRINGTABLE_DATA: &[u8; 4] = b"MMSD";
+pub const FILE_MAGIC_STRINGTABLE_INDEX: &[u8; 4] = b"MMSI";
+
+pub const FILE_HEADER_SIZE: usize = 8;
+
+pub fn write_file_header<S: SerializationSink>(s: &S, file_magic: &[u8; 4]) {
+    debug_assert_eq!(FILE_HEADER_SIZE, 8);
+
+    s.write_atomic(FILE_HEADER_SIZE, |bytes| {
+        bytes[0 .. 4].copy_from_slice(file_magic);
+        LittleEndian::write_u32(&mut bytes[4..8], CURRENT_FILE_FORMAT_VERSION);
+    });
+}
+
+pub fn read_file_header(
+    bytes: &[u8],
+    expected_magic: &[u8; 4]
+) -> Result<u32, Box<dyn Error>> {
+    debug_assert_eq!(FILE_HEADER_SIZE, 8);
```

Comment:
I'm a bit confused by these asserts -- they don't need to be debug as I expect LLVM to constant-fold them away, and either way I'm not sure what this is intended to do (i.e., because we're comparing against a constant declared in this file). Maybe a comment on the constant that these functions need updating would be better?

Reply:
The assertions are meant to be an extra safeguard against accidentally changing something here. I'd like to leave them in (since it's easy to overlook a comment but not so easy to overlook CI going red). I'll add some comments explaining their purpose though.